backend/src/common/types/controller/controller.ts

```diff
@@ -1,26 +1,32 @@
+import { TokenPayload } from '@autoline/shared';
 import { Request } from 'express';
 import { Query, ParamsDictionary } from 'express-serve-static-core';
 
-interface TypedRequestBody<T> extends Request {
+interface TypedRequestBody<T> extends AuthRequest {
   body: T;
 }
 
-interface TypedRequestQuery<T extends Query> extends Request {
+interface TypedRequestQuery<T extends Query> extends AuthRequest {
   query: T;
 }
 
-interface TypedRequestParams<T extends ParamsDictionary> extends Request {
+interface TypedRequestParams<T extends ParamsDictionary> extends AuthRequest {
   params: T;
 }
 
-interface TypedRequest<T extends Query, U> extends Request {
+interface TypedRequest<T extends Query, U> extends AuthRequest {
   body: U;
   query: T;
 }
 
+interface AuthRequest extends Request {
+  tokenPayload?: TokenPayload;
+}
```

Comment:
This can be done in a simpler way and w/o `as`:

```ts
type AuthTypedRequestBody<T> = TypedRequestBody<T> & AuthRequest;

type AuthTypedRequestQuery<T extends Query> = TypedRequestQuery<T> &
  AuthRequest;

type AuthTypedRequestParams<T extends ParamsDictionary> =
  TypedRequestParams<T> & AuthRequest;

type AuthTypedRequestParams<T extends Query, U> = TypedRequest<T, U> &
  AuthRequest;

interface AuthRequest extends Request {
  tokenPayload: TokenPayload;
}
```

Reply:
We still have problems with types. If we use `req: AuthRequest` in role.middleware we will get TypeError in routers. I tried `Omit<>`, `Partial<>` and so on. Nothing helps but optional `tokenPayload?` in `AuthRequest`.

Maybe there is a way to specify one type as the input to middleware and another as the output?

Reply:
Where do you try to use the `req: AuthRequest`?

Reply:
In `role.middleware`. We get plain request and need to add tokenPayload to it